### DISCUSSION TITLE
Version 4

### DISCUSSION THREAD

**Discussion by andrasbacsai** [coollabsio/coolify#1027] [General]
Hello everyone! 👋🏻 

Maybe you already aware that I started to work on the next generation of Coolify, called v4. It will be a complete rewrite from scratch. 

I want to level up Coolify to be more robust, be a no-brainer switch from other hosting platforms. For that, things needs to be changed from the core. I want to make Coolify to be more professional - it is not a side-project anymore, so I need to think about it differently.

## Technical Details
I'm a Javascript guy. There is a severe problem with JS frameworks. I have worked on the backend and frontend for over eight years. So is JS good for Coolify? Yes, but… 
The DX is not the best. The frameworks are limited and only do one or a few things well. They often sacrifice something for the greater good and try to solve a problem I do not have.
(I have an upcoming announcement for this. I really would like to solve this problem in the Javascript world.

I 💜 JS, but I realized that I love DX even more during the past months. When it just works, it is a magical feeling. So yeah. I decided to switch to a framework where I could develop even faster, smoother, and produce simpler code. When I decided to take this move, I feared what the community would think, but I quickly realized that you probably don't care. All you care about is a great application. (Right? 😃 )

Without further ado, it will be **Laravel**. Yes. PHP. In 2023. I've spent the last few weeks learning every aspect and love it. 

## What is this discussion about?
I already started to discuss some details with you on [Discord](https://coollabs.io/discord), but I would like to include people from GitHub as well.

**Please share your thoughts about what features/improvements should be included in v4**

**I already have a plan (soon in written form), but as you are the real users, I would like to include you in the process as much as I could.**

> But please, do not try to convince me about that I should use JS, Python, Rust or anything else instead of PHP. 😆 

## FAQ
(Will be extended continuously)

Q: Is v3 upgradable to v4?
A: Not in a straightforward way (just by clicking on the update button on Coolify’s UI), but I will add an option to import your resources from v3 to a v4 instance.I try to make it the most simple and easy task, without any downtime. I will keep you updated! There will be a migration path available.

Q: Any release date?
A: Not yet. I will share if I have one, please do not ask.

Q: Is v3 will be deprecated after v4 is released?
A: For a while no. I won't add new features to it, only bugfixes.

Q: Will it be based on SvelteKit?
A: I love SvelteKit, but unfortunately it is just a frontend framework, so it is not the best thing for Coolify. I definitely still use it for other things!

Q: v4 will be Apache licensed?
A: Most probably, yes. If you have an idea why to use different one, let me know.

Q: Is there already a public repo?
A: Yes, it is located in the v4 branch of [this](https://github.com/coollabsio/coolify/tree/v4) repository.

Q: Do you have a clear roadmap / issues so we can see how we're able to contribute?
A: Yes, but not in written form. I will dump it from my head soon.

**Comment by Wulfheart** [comment]
Hey,
what features are going to be in v4?

**Comment by Fronix** [comment]
Not sure where to put suggesions so I'll just type here

I'd like for V4 to have official support to use your own traefik instance, the current way of hacking it into coolify doesn't sit right with me.

**Comment by kazzkiq** [comment]
Just learned about Coolify and fell into this thread.

While I believe no language is silver bullet, and while you make clear that you won't be switching from your PHP choice, I feel the need just expose some thoughts:

There is this concern that PaaS should consume the minimal amount of resources as possible from your machine. The reason is quite obvious: If you're running your PaaS along with your apps, every single byte not used by the PaaS is available to your apps instead.

Neither JavaScript nor PHP are known as the best options for optimal usage of machine resources. There are better alternatives like Go and Rust, and in fact, many Docker-related projects use Go because of this.

But you're right when you say DX is also important. Rust is awesome for performance, but is the opposite for DX. Go is way easier than Rust but also not as easy as PHP + Laravel.

Perhaps a way to keep Laravel DX without the concern of Laravel/PHP performance and resource usage ([which can be poor](https://www.techempower.com/benchmarks/#section=data-r21&l=yykgsf-6bj&test=fortune&f=zik0zj-zik073-zik0zj-zik0zj-zik0lb-zik0zj-zih7un-zieeov-zik0zj-zik0xr-zik0zj-zik0zj-zik0zj-35r)) is to **implement cluster-mode and/or fully detach the build machine from the services machine**.

By doing that, Coolify could make use of Laravel without caring about resource usage, since the apps would be running on other servers connected to the "build-server". This would be the main feature that, I believe, you should think when building v4, otherwise there is the risk of Coolify getting suspicious eyes from people who think PHP + Laravel was a bad choice resources-wise.

**Reply by Wulfheart** [reply]
These benchmark games do not matter in my opinion. Or at least they shouldn't.

**Comment by ChiefORZ** [comment]
I currently love the workflow with coolify v3, so in the ui not much has to change - it is really easy to understand if you used other PaaS like heroku or railway. Great work!
One addition i would love is a Queue Overview for the running build jobs. Not only do i want to see the order of the Build queue, but it would be awesome to change the order/priority in the build queue and cancel jobs.
On top of that there is no option like "Wait for CI to pass before deploy". When my CI runner fails, the Automatic Deployment should not be executed.

Other than that just small improvements, like exporting all secrets of an application to json or .env files.
